src/contracts/sgho/sGHO.sol

```diff
@@ -0,0 +1,380 @@
+    uint256 currentAssets = totalAssets();
+    return currentAssets >= $.supplyCap ? 0 : $.supplyCap - currentAssets;
+  }
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function maxMint(address receiver) public view override returns (uint256) {
+    return convertToShares(maxDeposit(receiver));
+  }
+
+  /**
+   * @inheritdoc IsGHO
+   */
+  function depositWithPermit(
+    uint256 assets,
+    address receiver,
+    uint256 deadline,
+    SignatureParams memory sig
+  ) external returns (uint256) {
+    // Use permit to approve the vault to spend the user's GHO tokens
+    try
+      IERC20Permit(asset()).permit(
+        _msgSender(),
+        address(this),
+        assets,
+        deadline,
+        sig.v,
+        sig.r,
+        sig.s
+      )
+    {} catch {}
+
+    return deposit(assets, receiver);
+  }
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function totalAssets() public view override returns (uint256) {
+    return _convertToAssets(totalSupply(), Math.Rounding.Floor);
+  }
+
+  /// @inheritdoc IsGHO
+  function pause() external onlyRole(PAUSE_GUARDIAN_ROLE) {
+    _pause();
+  }
+
+  /// @inheritdoc IsGHO
+  function unpause() external onlyRole(PAUSE_GUARDIAN_ROLE) {
+    _unpause();
+  }
+
+  /**
+   * @inheritdoc IsGHO
+   */
+  function setTargetRate(uint16 newRate) public onlyRole(YIELD_MANAGER_ROLE) {
+    sGHOStorage storage $ = _getsGHOStorage();
+    // Update the yield index before changing the rate to ensure proper accrual
+    if (newRate > MAX_SAFE_RATE) {
+      revert RateMustBeLessThanMaxRate();
+    }
+    _updateYieldIndex();
+    $.targetRate = newRate;
+
+    // Convert targetRate from basis points to ray (1e27 scale)
+    // targetRate is in basis points (e.g., 1000 = 10%)
+    uint256 annualRateRay = (uint256(newRate) * RAY) / 10000;
+    // Calculate the rate per second (annual rate / seconds in a year)
+    $.ratePerSecond = (annualRateRay / 365 days).toUint96();
+
+    emit TargetRateUpdated(newRate);
+  }
+
+  /**
+   * @inheritdoc IsGHO
+   */
+  function setSupplyCap(uint160 newSupplyCap) public onlyRole(YIELD_MANAGER_ROLE) {
+    _getsGHOStorage().supplyCap = newSupplyCap;
+    emit SupplyCapUpdated(newSupplyCap);
+  }
+
+  /**
+   * @inheritdoc IRescuableBase
+   */
+  function maxRescue(
+    address erc20Token
+  ) public view override(IRescuableBase, RescuableBase) returns (uint256) {
+    if (erc20Token == asset()) {
+      return 0; // Cannot rescue GHO
+    }
+    return IERC20(erc20Token).balanceOf(address(this));
+  }
+
+  function _update(address from, address to, uint256 value) internal override whenNotPaused {
+    _updateYieldIndex();
+    super._update(from, to, value);
+  }
+
+  /**
+   * @notice Override _checkRescueGuardian to check for FUNDS_ADMIN role
+   * @dev This function reverts if the caller doesn't have the FUNDS_ADMIN role
+   */
+  function _checkRescueGuardian() internal view override {
+    if (!hasRole(FUNDS_ADMIN_ROLE, _msgSender())) {
```

Comment:
`FUNDS_ADMIN_ROLE` does not seem appropiate for the rescue role imho. The GSM uses `TOKEN_RESCUER_ROLE`, for reference.